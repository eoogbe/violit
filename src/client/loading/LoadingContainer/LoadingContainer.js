/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import Loading from 'react-loading';
import styled from 'styled-components';
import { Alert, theme } from 'client/core';
import useLoad from '../useLoad';

const StyledError = styled.div`
  font-size: 1rem;
  position: relative;
  height: 3.5em;
`;

const StyledLoading = styled(Loading)`
  font-size: 1rem;
  margin: ${({ theme }) => theme.spacingMd} auto;
`;

/** Holds data that needs to be loaded. */
const LoadingContainer = ({ loadData, action, render }) => {
  const [loading, error, data] = useLoad(loadData);

  if (error) {
    return (
      <StyledError>
        <Alert action={action} />
      </StyledError>
    );
  }

  if (loading) {
    return (
      <StyledLoading type="spinningBubbles" color={theme.primaryColor} />
    );
  }

  return render(data);
};

LoadingContainer.propTypes = {
  /** A message describing the attempted action. */
  action: PropTypes.string,
  /** Loads the data from the server. */
  loadData: PropTypes.func.isRequired,
  /** Renderer called when the server has returned data. */
  render: PropTypes.func.isRequired,
};

export default LoadingContainer;
