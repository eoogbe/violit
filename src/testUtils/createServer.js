/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import cookieParser from 'cookie-parser';
import express from 'express';
import routes from 'server/api/routes';

/** Creates a test server for the specified `router`. */
export default (router, locals) => {
  const app = express();
  app.use(cookieParser());
  app.use(express.json());
  app.use((req, res, next) => {
    Object.assign(res.locals, { routes }, locals);
    next();
  });
  app.use(router());
  app.use((err, req, res, next) => {
    res.status(500).json({ error: err.message });
  });
  return app;
};
