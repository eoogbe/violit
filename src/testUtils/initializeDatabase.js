/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import path from 'path';
import Database from 'better-sqlite3';
import bcrypt from 'bcrypt';
import * as BOARD from '__fixtures__/boardFixture';
import * as COMMENTS from '__fixtures__/commentsFixture';
import * as THREADS from '__fixtures__/threadsFixture';
import * as USER from '__fixtures__/userFixture';

const INSERT_USER_SQL =
`INSERT INTO user (id, username, password)
VALUES (:id, :username, :password)`;

const INSERT_BOARD_SQL =
`INSERT INTO board (id, name)
VALUES (:id, :name)`;

const INSERT_THREAD_SQL =
`INSERT INTO thread (
  id,
  slug,
  dateCreated,
  dateDeleted,
  headline,
  articleBody,
  userId,
  boardId
) VALUES (
  :id,
  :slug,
  :dateCreated,
  :dateDeleted,
  :headline,
  :articleBody,
  :userId,
  :boardId
)`;

const INSERT_COMMENT_SQL =
`INSERT INTO comment (id, dateCreated, dateDeleted, content, userId, threadId)
VALUES (:id, :dateCreated, :dateDeleted, :content, :userId, :threadId)`;

/** Initializes the test database. */
export default () => {
  const dbPath = path.resolve(__dirname, '..', '..', 'data', 'test.db');
  const db = new Database(dbPath, {
    fileMustExist: true,
  });

  db.prepare('DELETE FROM comment').run();
  db.prepare('DELETE FROM thread').run();
  db.prepare('DELETE FROM board').run();
  db.prepare('DELETE FROM user').run();

  db.prepare(INSERT_USER_SQL).run({
    id: 'user1',
    username: USER.username,
    password: bcrypt.hashSync(USER.password, 4),
  });

  db.prepare(INSERT_BOARD_SQL).run(BOARD.board);

  const insertThreadStmt = db.prepare(INSERT_THREAD_SQL);
  THREADS.db.forEach((thread) => {
    insertThreadStmt.run(thread);
  });

  const insertCommentStmt = db.prepare(INSERT_COMMENT_SQL);
  COMMENTS.db.forEach((comment) => {
    insertCommentStmt.run(comment);
  });

  return db;
};
