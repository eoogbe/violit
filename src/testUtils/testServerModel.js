/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as R from 'ramda';
import * as BOARD from '__fixtures__/boardFixture';
import * as COMMENTS from '__fixtures__/commentsFixture';
import * as THREADS from '__fixtures__/threadsFixture';
import * as USER from '__fixtures__/userFixture';

const CAN_DELETE_THREAD_RESULTS = {
  [THREADS.thread1.slug]: true,
  [THREADS.thread2.slug]: false,
  [THREADS.thread3.slug]: true,
};

const CAN_DELETE_COMMENT_RESULTS = {
  comment1: true,
  comment2: false,
  comment3: true,
};

/** Creates a test server model with default values. */
export const createModel = R.mergeDeepRight({
  auth: {
    sign: () => Promise.resolve('token123'),
    getCredentials: () => Promise.resolve(null),
    createCredentials: () => Promise.resolve('token123'),
  },
  board: {
    getBoard: () => BOARD.board,
  },
  threads: {
    getThreads: () => THREADS.unnormalized,
    getThreadById: (id) => ({
      id,
      slug: THREADS.thread1.slug,
      dateCreated: THREADS.thread1.dateCreated,
      headline: THREADS.thread1.headline,
      articleBody: THREADS.thread1.articleBody,
      author: THREADS.thread1.author,
    }),
    getThreadBySlug: (slug) => ({
      slug,
      id: 'thread1',
      dateCreated: THREADS.thread1.dateCreated,
      headline: THREADS.thread1.headline,
      articleBody: THREADS.thread1.articleBody,
      author: THREADS.thread1.author,
    }),
    createThread: () => THREADS.normalized.thread1,
    softDeleteThread: () => ({
      id: 'thread1',
      slug: THREADS.thread1.slug,
      dateCreated: THREADS.thread1.dateCreated,
      dateDeleted: THREADS.thread1.dateDeleted,
      headline: THREADS.thread1.headline,
      articleBody: THREADS.thread1.articleBody,
      author: THREADS.thread1.author,
    }),
    canDeleteThread: (thread) => CAN_DELETE_THREAD_RESULTS[thread.slug],
  },
  comments: {
    getCommentsByThreadSlug: () => COMMENTS.unnormalized,
    getCommentById: (id) => ({
      ...COMMENTS.normalized.comment1,
      id,
    }),
    createComment: () => COMMENTS.normalized.comment1,
    softDeleteComment: (id) => ({
      id,
      dateCreated: COMMENTS.comment1.dateCreated,
      dateDeleted: COMMENTS.comment1.dateDeleted,
      text: COMMENTS.comment1.text,
      threadSlug: COMMENTS.comment1.threadSlug,
      author: COMMENTS.comment1.author,
    }),
    canDeleteComment: (comment) => CAN_DELETE_COMMENT_RESULTS[comment.id],
  },
  users: {
    getUserById: () => USER.credentials,
    getUserByUsername: (username) => ({
      ...USER.credentials,
      username,
    }),
    createUser: () => Promise.resolve(USER.credentials),
  },
});

/** Creates a test server model for an authenticated user. */
export const createAuthenticatedModel =
  R.mergeDeepRight(createModel({
    auth: {
      getCredentials: () => Promise.resolve(USER.credentials),
    },
  }));
