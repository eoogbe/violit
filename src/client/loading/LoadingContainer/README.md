```jsx
const loadData = () => Promise.resolve('foo');
const render = (data) => <div>{data}</div>;

<LoadingContainer loadData={loadData} render={render} />
```

When loading (refresh page if loaded):

```jsx
import { wait } from 'testUtils/wait';

const loadData = () => wait(1000000);
const render = (data) => <div>{data}</div>;

<LoadingContainer loadData={loadData} render={render} />
```

When error:

```jsx
const loadData = () => Promise.reject();
const render = (data) => <div>{data}</div>;

<LoadingContainer
  loadData={loadData}
  action="loading the foo"
  render={render}
/>
```
