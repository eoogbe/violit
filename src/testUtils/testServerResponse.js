/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as R from 'ramda';
import routes from 'server/api/routes';
import { createAuthenticatedModel, createModel } from './testServerModel';
import * as USER from '__fixtures__/userFixture';

/** Creates a test response object with default values. */
export const createResponse = R.mergeDeepRight({
  status: jest.fn().mockReturnThis(),
  set: jest.fn().mockReturnThis(),
  clearCookie: jest.fn().mockReturnThis(),
  send: jest.fn(),
  locals: {
    model: createModel({}),
    routes,
  },
});

/** Creates a test response object for an authenticated user. */
export const createAuthenticatedResponse = R.mergeDeepRight(createResponse({
  locals: {
    model: createAuthenticatedModel({}),
    credentials: USER.credentials,
  },
}));
