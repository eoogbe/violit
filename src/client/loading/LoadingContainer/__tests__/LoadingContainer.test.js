/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import TestRenderer from 'react-test-renderer';
import { wait } from 'testUtils/wait';
import LoadingContainer from '../LoadingContainer';

const ACTION = 'loading the foo';

it('renders without crashing', async () => {
  const loadData = () => Promise.resolve('foo');
  const render = (data) => <div>{data}</div>;

  let renderer;
  await TestRenderer.act(async () => {
    renderer = TestRenderer.create(
      <LoadingContainer loadData={loadData} action={ACTION} render={render} />
    );

    await wait();
  });

  expect(renderer.toJSON()).toMatchSnapshot();
});

it('renders when loading', async () => {
  const loadData = () => wait(5);
  const render = (data) => <div>{data}</div>;

  let renderer;
  await TestRenderer.act(async () => {
    renderer = TestRenderer.create(
      <LoadingContainer loadData={loadData} action={ACTION} render={render} />
    );
  });

  expect(renderer.toJSON()).toMatchSnapshot();
});

it('renders with error', async () => {
  const loadData = () => Promise.reject();
  const render = (data) => <div>{data}</div>;

  let renderer;
  await TestRenderer.act(async () => {
    renderer = TestRenderer.create(
      <LoadingContainer loadData={loadData} action={ACTION} render={render} />
    );

    await wait();
  });

  expect(renderer.toJSON()).toMatchSnapshot();
});
