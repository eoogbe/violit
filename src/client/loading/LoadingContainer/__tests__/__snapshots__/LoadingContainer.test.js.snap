// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when loading 1`] = `
<div
  className="sc-htoDjs eYXpaL"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\">
  <circle cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(45 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.125s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(90 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.25s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(135 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.375s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(180 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.5s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(225 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.625s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(270 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.75s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(315 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.875s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(180 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.5s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
</svg>
",
    }
  }
  style={
    Object {
      "fill": "#7F00FF",
      "height": 64,
      "width": 64,
    }
  }
/>
`;

exports[`renders with error 1`] = `
<div
  className="sc-gzVnrw fPXbEJ"
>
  <div
    aria-live="assertive"
    className="sc-bdVaJa dXBAhf"
    role="alert"
  >
    <p
      className="sc-htpNat bhpHJD"
    >
      Oops! Something went wrong with 
      loading the foo
      !Â 
      <button
        className="sc-bxivhb ErYiX"
        onClick={[Function]}
        type="button"
      >
        Reload page
      </button>
    </p>
    <button
      className="sc-bwzfXH haUZkm"
      onClick={[Function]}
      type="button"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="1.25em"
        size="1.25em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        title="Dismiss alert"
        viewBox="0 0 24 24"
        width="1.25em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`renders without crashing 1`] = `
<div>
  foo
</div>
`;
